src/graph/heterograph.cc

```diff
@@ -260,6 +261,144 @@ HeteroGraphPtr HeteroGraph::CopyTo(HeteroGraphPtr g, const DLContext& ctx) {
                                         hgindex->num_verts_per_type_));
 }
 
+std::string HeteroGraph::SharedMemName() const {
+  return shared_mem_ ? shared_mem_->GetName() : "";
+}
+
+HeteroGraphPtr HeteroGraph::CopyToSharedMem(
+      HeteroGraphPtr g, const std::string& name, const std::set<std::string>& fmts) {
+  auto hg = std::dynamic_pointer_cast<HeteroGraph>(g);
+  if (hg->SharedMemName() == name)
+    return g;
+
+  // Copy buffer to share memory
+  auto mem = std::make_shared<SharedMemory>(name);
+  auto mem_buf = mem->CreateNew(SHARED_MEM_METAINFO_SIZE_MAX);
+  dmlc::MemoryFixedSizeStream ofs(mem_buf, SHARED_MEM_METAINFO_SIZE_MAX);
+  dmlc::SeekStream *strm = &ofs;
+
+  bool has_coo = fmts.find("coo") != fmts.end();
+  bool has_csr = fmts.find("csr") != fmts.end();
+  bool has_csc = fmts.find("csc") != fmts.end();
+  strm->Write(g->NumBits());
+  strm->Write(has_coo);
+  strm->Write(has_csr);
+  strm->Write(has_csc);
+  strm->Write(ImmutableGraph::ToImmutable(hg->meta_graph_));
+  strm->Write(hg->num_verts_per_type_);
+
+  std::vector<HeteroGraphPtr> relgraphs(g->NumEdgeTypes());
+
+  for (dgl_type_t etype = 0 ; etype < g->NumEdgeTypes() ; ++etype) {
+    strm->Write(hg->NumEdges(etype));
+    aten::COOMatrix coo;
+    aten::CSRMatrix csr, csc;
+    std::string prefix = name + "_" + std::to_string(etype);
+    if (has_coo) {
+      coo = hg->GetCOOMatrix(etype).CopyToSharedMem(prefix + "_coo");
+      strm->Write(coo.row_sorted);
+      strm->Write(coo.col_sorted);
+    }
+    if (has_csr) {
+      csr = hg->GetCSRMatrix(etype).CopyToSharedMem(prefix + "_csr");
+      strm->Write(csr.sorted);
+    }
+    if (has_csc) {
+      csc = hg->GetCSCMatrix(etype).CopyToSharedMem(prefix + "_csc");
+      strm->Write(csc.sorted);
+    }
+    relgraphs[etype] = UnitGraph::CreateHomographFrom(csc, csr, coo, has_csc, has_csr, has_coo);
+  }
+
+  auto ret = HeteroGraphPtr(new HeteroGraph(hg->meta_graph_, relgraphs, hg->num_verts_per_type_));
+  auto hg_index = std::dynamic_pointer_cast<HeteroGraph>(ret);
```

Comment:
According to https://stackoverflow.com/a/46836519, seems you can directly return shared_ptr<HeteroGraph> as HeteroGraphPtr. Could you try it? dynamic_pointer_cast is not so safe as downcasting

Reply:
Here I need to modify the returned HeteroGraph object so I do this dynamic_pointer_cast.

Reply:
What I mean is to directly `std::shared_ptr<HeteroGraph>(new HeteroGraph(...))` and `std::static_pointer_cast<BaseHeteroGraph>(...)` to the returned object if necessary.